### ISSUE TITLE
Amplify analytics pinpoint throw exception on Android

### ISSUE THREAD

**Issue by JakubWijata** [aws-amplify/amplify-flutter#3106]
### Description

After migration to amplify 1.1.0 from 0.6.12 pinpoint starts throwing "Forbidden from accessing the Pinpoint resource" exception on Android. The problem exists only on Android and occurred on the new version of Amplify. On ios works well.

 ```E/flutter (10506): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: UnknownException {
E/flutter (10506):   "message": "Forbidden from accessing the Pinpoint resource.",
E/flutter (10506):   "recoverySuggestion": "Please ensure Analytics is properly configured in your amplifyconfiguration.dart file and that unauthenticated guest access is allowed or an authorized user is signed in.",
E/flutter (10506):   "underlyingException": "ForbiddenException {\n}"
E/flutter (10506): }
E/flutter (10506): #0      EndpointClient.updateEndpoint
package:amplify_analytics_pinpoint_dart/…/endpoint_client/endpoint_client.dart:172
E/flutter (10506): <asynchronous suspension>
E/flutter (10506):

```

### Categories

- [X] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [ ] Auth
- [ ] Authenticator
- [ ] DataStore
- [ ] Storage

### Platforms

- [ ] iOS
- [X] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.10.2

### Amplify Flutter Version

1.1.0

### Deployment Method

Amplify CLI

**Comment by Jordan-Nelson** [comment]
@JakubWijata can you share your amplify config (you can remove any sensitive data or app specific data such as API keys or app ids).

Does the issue persist if you uninstall and reinstall the app?

During first app launch after upgrading from 0.x to 1.x there is some on device data migrated. I'm wondering if the exception could be related. If it happens after uninstall/re-install it that is likely not related.

**Comment by jamontesg** [comment]
I have a similar error .
I only has    a basic configuration :

static configureAmplify() async {
    if (!zIsWeb && (Platform.isAndroid || Platform.isIOS)) {
      AmplifyDataStore datastorePlugin = AmplifyDataStore(
        modelProvider: ModelProvider.instance,
        errorHandler: ((error) =>
            {developer.log("Custom ErrorHandler received DS: $error")}),
      );

      await Amplify.addPlugin(datastorePlugin);
    }
    AmplifyAuthCognito authPlugin = AmplifyAuthCognito();
    await Amplify.addPlugin(authPlugin);

    AmplifyAPI apiPlugin = AmplifyAPI();
    await Amplify.addPlugin(apiPlugin);

    AmplifyStorageS3 storagePlugin = AmplifyStorageS3();
    await Amplify.addPlugin(storagePlugin);

    AmplifyAnalyticsPinpoint pintPointPlugin = AmplifyAnalyticsPinpoint();
    await Amplify.addPlugin(pintPointPlugin);
    Amplify.Hub.listen(HubChannel.Auth, (hubEvent) async {
      //Amplify.Hub.listen(HubChannel.Auth, (dynamic hubEvent) async {
      switch (hubEvent.eventName) {
        case 'SIGNED_IN':
          developer.log('USER IS SIGNED IN');
          break;
        case 'SIGNED_OUT':
          //print('USER IS SIGNED OUT');
          break;
        case 'SESSION_EXPIRED':
          //print('SESSION HAS EXPIRED');
          break;
        case 'USER_DELETED':
          //print('USER HAS BEEN DELETED');
          break;
        default:
          developer.log('DEFAULT: ${hubEvent.eventName}');
          break;
      }
    });
    
    try {
      if (Amplify.isConfigured) {
        await Amplify.configure(amplifyconfig);
      } else {
        await Amplify.configure(amplifyconfig);
      }
    } on Exception catch (e) {
      safePrint('An error occurred configuring Amplifys--> : $e');
    }





and my error  is:


ERROR | EventClient | Unrecoverable issue, deleting cache for local event batch
E/flutter (29268): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: UnknownException {
E/flutter (29268):   "message": "Forbidden from accessing the Pinpoint resource.",
E/flutter (29268):   "recoverySuggestion": "Please ensure Analytics is properly configured in your amplifyconfiguration.dart file and that unauthenticated guest access is allowed or an authorized user is signed in.",
E/flutter (29268):   "underlyingException": "ForbiddenException {\n}"
E/flutter (29268): }

**Comment by JakubWijata** [comment]
> @JakubWijata can you share your amplify config (you can remove any sensitive data or app specific data such as API keys or app ids).
> 
> Does the issue persist if you uninstall and reinstall the app?
> 
> During first app launch after upgrading from 0.x to 1.x there is some on device data migrated. I'm wondering if the exception could be related. If it happens after uninstall/re-install it that is likely not related.

@Jordan-Nelson Yes. This issue persists even if I reinstall the app from scratch.

here is my config:

```{
  "UserAgent": "aws-amplify-cli/2.0",
  "Version": "1.0",
  "auth": {
    "plugins": {
      "awsCognitoAuthPlugin": {
        "UserAgent": "aws-amplify-cli/0.1.0",
        "Version": "0.1.0",
        "IdentityManager": {"Default": {}},
        "CredentialsProvider": {
          "CognitoIdentity": {
            "Default": {
              "PoolId": "********",
              "Region": "********"
            }
          }
        },
        "CognitoUserPool": {
          "Default": {
            "PoolId": "********",
            "AppClientId": "********",
            "AppClientSecret":
                "********",
            "Region": "********"
          }
        },
        "Auth": {
          "Default": {"authenticationFlowType": "USER_SRP_AUTH"}
        }
      }
    }
  },
  "analytics": {
    "plugins": {
      "awsPinpointAnalyticsPlugin": {
        "pinpointAnalytics": {
          "appId": "********",
          "region": "********"
        },
        "pinpointTargeting": {"region": "********"},
        "autoFlushEventsInterval": 10000
      }
    }
  },
  "storage": {
    "plugins": {
      "awsS3StoragePlugin": {
        "bucket": "********",
        "region": "********"
      }
    }
  }
}```

**Comment by MatijaOsojnik** [comment]
Experiencing the same issue - I've just added the `analytics` through `amplify add analytics` using default configurations settings.

**Comment by Jordan-Nelson** [comment]
I have not been able to reproduce this issue. I have a few follow up questions for those facing this issue.

1. Did you use the Amplify CLI to create the AWS resources or were they created via some other means (CDK, AWS console, etc.)
2. Have you enabled analytics events to be sent for both authorized and unauthorized users, or only for authorized users?
3. @jamontesg @MatijaOsojnik - Were you previously using 0.6.x successfully?

**Comment by JakubWijata** [comment]
2. Only for authorized users but I followed the doc and after log in using Amplify auth I use analytics identifyUser method. But as I said before amplify update works correctly  
```
    final userProfile = AnalyticsUserProfile();
    final properties = AnalyticsProperties();
    final currentUser = await Amplify.Auth.getCurrentUser();

    properties.addStringProperty(propertyUserSub, session.userSub!);
    userProfile.properties = properties;

    await Amplify.Analytics.identifyUser(
      userId: currentUser.userId,
      userProfile: userProfile,
    );
```



> I have not been able to reproduce this issue. I have a few follow up questions for those facing this issue.
> 
> 1. Did you use the Amplify CLI to create the AWS resources or were they created via some other means (CDK, AWS console, etc.)
> 2. Have you enabled analytics events to be sent for both authorized and unauthorized users, or only for authorized users?
> 3. @jamontesg @MatijaOsojnik - Were you previously using 0.6.x successfully?

**Comment by jamontesg** [comment]
Hi I am using cli, and  only  has  plugin initialization. nothing more

**Comment by dJani97** [comment]
We have the same issue, once a user logs in, Amplify fails to configure and the app doesn't start anymore. App data has to be cleared for it to work again.

Note that this only happens when a User is added to a Cognito group. If the user has no group, the error goes away.

We're on the latest Flutter and Amplify.

**Comment by fjnoyp** [comment]
Hi everyone, thank you for sharing the details of your issues here to help us debug. 

The issue might be Analytics code attempting to make authenticated calls when it's configuring.  Since an auth login is required, the call fails and puts Analytics in a bad state.  

Looking to reproduce this to confirm.  Thanks all for your patience here.

**Comment by JakubWijata** [comment]
updating flutter to 3.10.3 fixed it for me

**Comment by dJani97** [comment]
I just tried Flutter `3.10.3` and Amplify `1.1.1`, and it didn't fix the problem for me.

**Comment by fjnoyp** [comment]
#### I Cannot Reproduce Your Errors
With Amplify 1.1.1 or 1.1.0, with Analytics guest access disabled, I do not get any **ForbiddenException**.  I'm using Flutter 3.11.0 and Dart 3.1.0 

#### SessionExpiredException/NotAuthorizedException
I get these exceptions when disabling Analytics guest access on all platforms.  As it seems unrelated to your issues, I created a new one here: https://github.com/aws-amplify/amplify-flutter/issues/3144

## **My Responses to your Issues:**

#### @JakubWijata
Good to hear.  Amplify `1.1.1` modified the Amplify Auth Android code and the auth migration code (for 0.6 -> 1.0.0 upgrades) which might have fixed your issue. 

#### @jamontesg & @MatijaOsojnik 
Can you do what @JakubWijata did and upgrade to Flutter > `3.10.3` and Amplify `1.1.1`? 

#### @dJani97 
1. Sorry if too simple, but you ran `flutter clean` and `flutter pub get` after upgrading right?  
2. Did you use Amplify CLI to setup your project? 
3. What Cognito Group are you adding the user to?  I manually added my user to a custom Cognito Group and didn't observe any issue. 
4. Can you confirm you are getting the `Forbidden from accessing the Pinpoint resource` exception?  Otherwise, you might have a different issue here.

**Comment by dJani97** [comment]
@fjnoyp yes, we used Amplify CLI to setup & add push notification support to the project.

I think we figured out a solution. There is an IAM role associated with each Cognito group, and we had to add the following Statement to these roles:

```json
"Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "mobiletargeting:*",
                "mobileanalytics:*"
            ],
            "Resource": "*"
        }
    ]
```

Isn't this something that should be set up automatically by Amplify CLI?

Source: [https://serverfault.com/questions/875627/aws-iam-user-credentials-for-accessing-pinpoint-through-script-causes-internals](https://serverfault.com/questions/875627/aws-iam-user-credentials-for-accessing-pinpoint-through-script-causes-internals)

Is this a good solution? Allowing `*` seems a bit overkill, should we narrow this down to something?

**Comment by jamontesg** [comment]
@fjnoyp  Same error after upgrade:
Flutter version 3.10.3
amplify_analytics_pinpoint: ^1.1.1

Using amplify cli , cognito and  auth by groups


ERROR | EventClient | Unrecoverable issue, deleting cache for local event batch
E/flutter (16880): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: UnknownException {
E/flutter (16880):   "message": "Forbidden from accessing the Pinpoint resource.",
E/flutter (16880):   "recoverySuggestion": "Please ensure Analytics is properly configured in your amplifyconfiguration.dart file and that unauthenticated guest access is allowed or an authorized user is signed in.",
E/flutter (16880):   "underlyingException": "ForbiddenException {\n}"
E/flutter (16880): }

**Comment by fjnoyp** [comment]
@dJani97 
Thanks for sharing that update and glad to hear it's working for you. 

Amplify CLI automatically sets up the necessary IAM role permissions for the "Authenticated role" in your Identity Pool.   You can find yours by going to IAM -> Roles, and searching by your app name.  Open the `amplify-<your app name>-authRole` role to see.

Do you have multiple Identity Pools or a special Auth setup?  If you share your `amplifyconfiguration.dart` file with keys removed I can get maybe get a better idea of what's going on. 

And no that's not a good solution as it gives access to all resources.  You can edit the IAM role in the `Visual editor`, and click on 'Add ARN' to specify the proper Pinpoint resource:

![image](https://github.com/aws-amplify/amplify-flutter/assets/11478656/294a93b8-3b3e-4bef-8a57-2b653e288480)

**Comment by fjnoyp** [comment]
@dJani97 
Thanks for sharing that update and glad to hear it's working for you. 

### Amplify CLI Auth Setup 
Amplify CLI automatically sets up the necessary IAM role permissions for the "Authenticated role" in your Identity Pool.   You can find yours by going to IAM -> Roles, and searching by your app name.  Open the `amplify-<your app name>-authRole` role to see.  That can be a good basis for knowing how to update your IAM roles.  

### Using * in IAM Role
No that's not a good solution as it gives access to all resources.  You can edit the IAM role in the `Visual editor`, and click on 'Add ARN' to specify the proper Pinpoint resource:

![image](https://github.com/aws-amplify/amplify-flutter/assets/11478656/294a93b8-3b3e-4bef-8a57-2b653e288480)

**Comment by fjnoyp** [comment]
@jamontesg 

It sounds like you might have the same issue as **dJani97** 

Can you please make sure that the IAM roles attached to your User Pool Group has proper permission to access Pinpoint?  If you are unsure how to do that, please see my comment above to **dJani97**

**Comment by jamontesg** [comment]
Hi @fjnoyp , using idea  of @dJani97  works fine.  for me this is a bug. pinpoint  only works with  basic authentication.

**Comment by windcludy** [comment]
You can solve this problem by add a customPolicies to auth/user-pool-group-precedence.json .

see the following snippet for an example auth/user-pool-group-precedence.json file.

[
  {
    "groupName": "SuperAdmins",
    "precedence": 1,
    "customPolicies": [
      {
        "PolicyName": "pinpoint-put",
        "PolicyDocument": {
          "Version": "2012-10-17",
          "Statement": [
            {
              "Effect": "Allow",
              "Action": [
                "mobiletargeting:UpdateEndpoint",
                "mobiletargeting:PutEvents"
              ],
              "Resource": [
                "arn:aws:mobiletargeting:*:YOU_ORG_ID:apps/YOUR_APP_ID*"
              ]
            }
          ]
        }
      }
    ]
  },
  {
    "groupName": "Admins",
    "precedence": 2,
    "customPolicies": [
      {
        "PolicyName": "pinpoint-put",
        "PolicyDocument": {
          "Version": "2012-10-17",
          "Statement": [
            {
              "Effect": "Allow",
              "Action": [
                "mobiletargeting:UpdateEndpoint",
                "mobiletargeting:PutEvents"
              ],
              "Resource": [
                "arn:aws:mobiletargeting:*:YOUR_ORG_ID:apps/YOUR_APP_ID*"
              ]
            }
          ]
        }
      }
    ]
  }
]

**Comment by jamontesg** [comment]
@fjnoyp access guest disabled ,  amplify cli 12.0.3.   for group asignation I use post confirmation trigger (lambda).
Thanks  for your time.

**Comment by krunalSutariya** [comment]
I'm getting the same error.. 

Code:-

  await Amplify.addPlugins(
          [auth, storage, apiPlugin, dataStorePlugin, analytics]);
      print(Amplify.isConfigured);
      await Amplify.configure(amplifyconfig);
      
      
  Error:- Exception has occurred.
UnknownException (UnknownException {
  "message": "Forbidden from accessing the Pinpoint resource.",
  "recoverySuggestion": "Please ensure Analytics is properly configured in your amplifyconfiguration.dart file and that unauthenticated guest access is allowed or an authorized user is signed in.",
  "underlyingException": "ForbiddenException {\n}"
})

**Comment by likeconan** [comment]
Any updates for this issue? The amplify configuration in ios is working but not in android.

**Comment by dgagnon** [comment]
@windcludy doesn't this policy allows all endpoints to update all endpoints ?  As it, I can grab the token and push directly to pinpoint with a manufactured payload and modify any endpoint.

```
mobiletargeting:UpdateEndpoint
Create an endpoint or update the information for an endpoint.

URI – [/apps/projectId/endpoints/endpointId](https://docs.aws.amazon.com/pinpoint/latest/apireference/rest-api-endpoint.html#rest-api-endpoint-methods-put)

Method – PUT

Resource ARN – arn:aws:mobiletargeting:region:accountId:apps/projectId/endpoints/endpointId
```

**Comment by dnys1** [comment]
The fix for the original issue (being able to not catch the unhandled exception) was released in Amplify v1.2.1.

For those experiencing issues related to integration with user pool groups, see [this comment](https://github.com/aws-amplify/amplify-flutter/issues/3106#issuecomment-1576938954) and [this comment](https://github.com/aws-amplify/amplify-flutter/issues/3106#issuecomment-1581926493). Although this was [fixed](https://github.com/aws-amplify/amplify-cli/pull/10455) in the CLI some time ago, so I would recommend opening an issue in the [amplify-cli](https://github.com/aws-amplify/amplify-cli/) repo if a newer CLI version does not resolve the issue.

For those that are still experiencing issues, please open a new ticket with the necessary information to help us investigate.
